ios/RNRadar.m

```diff
@@ -88,79 +111,539 @@ - (void)didFailWithStatus:(RadarStatus)status {
+}
+
+RCT_EXPORT_METHOD(startTrackingContinuous) {
+    [Radar startTrackingWithOptions:RadarTrackingOptions.continuous];
+}
+
+RCT_EXPORT_METHOD(startTrackingCustom:(NSDictionary *)optionsDict) {
+    RadarTrackingOptions *options = [RadarTrackingOptions trackingOptionsFromDictionary:optionsDict];
+    [Radar startTrackingWithOptions:options];
+}
+
+RCT_EXPORT_METHOD(stopTracking) {
+    [Radar stopTracking];
+}
+
+RCT_EXPORT_METHOD(acceptEvent:(NSString *)eventId verifiedPlaceId:(NSString *)verifiedPlaceId) {
+    [Radar acceptEventId:eventId verifiedPlaceId:verifiedPlaceId];
 }
 
 RCT_EXPORT_METHOD(rejectEvent:(NSString *)eventId) {
     [Radar rejectEventId:eventId];
 }
 
+RCT_EXPORT_METHOD(getContext:(NSDictionary *)locationDict resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject) {
+    CLLocation *location;
+    if (locationDict != nil && [locationDict isKindOfClass:[NSDictionary class]]) {
+        double latitude = [RCTConvert double:locationDict[@"latitude"]];
+        double longitude = [RCTConvert double:locationDict[@"longitude"]];
+        NSDate *timestamp = [NSDate new];
+        location = [[CLLocation alloc] initWithCoordinate:CLLocationCoordinate2DMake(latitude, longitude) altitude:-1 horizontalAccuracy:5 verticalAccuracy:-1 timestamp:timestamp];
+    }
+
+    __block RCTPromiseResolveBlock resolver = resolve;
+    __block RCTPromiseRejectBlock rejecter = reject;
+
+    RadarContextCompletionHandler completionHandler = ^(RadarStatus status, CLLocation * _Nullable location, RadarContext * _Nullable context) {
+        if (status == RadarStatusSuccess && resolver) {
+            NSMutableDictionary *dict = [NSMutableDictionary new];
+            [dict setObject:[Radar stringForStatus:status] forKey:@"status"];
+            if (location) {
+                [dict setObject:[Radar dictionaryForLocation:location] forKey:@"location"];
+            }
+            if (context) {
+                [dict setObject:[context dictionaryValue] forKey:@"context"];
+            }
+            resolver(dict);
+        } else if (rejecter) {
+            rejecter([Radar stringForStatus:status], [Radar stringForStatus:status], nil);
+        }
+        resolver = nil;
+        rejecter = nil;
+    };
+
+    if (location) {
+        [Radar getContextForLocation:location completionHandler:completionHandler];
+    } else {
+        [Radar getContextWithCompletionHandler:completionHandler];
+    }
+}
+
+RCT_EXPORT_METHOD(searchPlaces:(NSDictionary *)optionsDict resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject) {
+    if (optionsDict == nil) {
+        if (reject) {
+            reject([Radar stringForStatus:RadarStatusErrorBadRequest], [Radar stringForStatus:RadarStatusErrorBadRequest], nil);
+        }
+
+        return;
+    }
+
+    NSDictionary *nearDict = optionsDict[@"near"];
+    CLLocation *near;
+    if (nearDict != nil && [nearDict isKindOfClass:[NSDictionary class]]) {
```

Comment:
looks like `near` is required on the server - https://github.com/radarlabs/server/blob/master/app.ts#L565

do we want to do any upstream validation or let it 400 over the wire?

Reply:
Look a little bit lower. If `near` is provided, we call `searchPlacesNear:` (provided location), otherwise we call `searchPlacesWithRadius:` (get current device location)